src/Cms/Binding.php

```diff
@@ -0,0 +1,417 @@
+			$this->create();
+		}
+
+		return $this->protocol->tostring(false);
+	}
+
+	/**
+	 * Populates cache with UUIDs
+	 * for all pages and files in index
+	 *
+	 * @return void
+	 */
+	public static function index(): void
+	{
+		$kirby = App::instance();
+		$site  = $kirby->site();
+		$index = $site->index(true);
+
+		// pages
+		foreach ($index as $page) {
+			static::for($page)->populate();
+		}
+
+		// page files
+		foreach ($index->files() as $file) {
+			static::for($file)->populate();
+		}
+
+		// site files
+		foreach ($site->files() as $file) {
+			static::for($file)->populate();
+		}
+
+		// user files
+		foreach ($kirby->users()->files() as $file) {
+			static::for($file)->populate();
+		}
+
+		// TODO: finding all structure fields and adding UUIDs
+		// blocks already have a UUID id
+	}
+
+	/**
+	 * Returns key used for cache
+	 *
+	 * @return string
+	 */
+	public function key(): string
+	{
+		$id = $this->protocol->toString(false);
+
+		// for better performance when using a file-based cache,
+		// turn first two characters of the id into a directory
+		$id = substr($id, 0, 2) . '/' . substr($id, 2);
+
+		// TODO: if we also do blocks and structure fields, maybe good to
+		// add an optional second level of directories?
+
+		return $this->type() . '/' . $id;
+	}
+
+	/**
+	 * Write UUID binding to cache
+	 *
+	 * @return bool
+	 */
+	public function populate(): bool
+	{
+		return static::cache()->set($this->key(), $this->value());
+	}
+
+	/**
+	 * Resolve binding from UUID to site/page/file/user
+	 *
+	 * @return \Kirby\Cms\Page|\Kirby\Cms\File|\Kirby\Cms\User|\Kirby\Cms\Site|null
+	 */
+	public function model(): Page|File|User|Site|null
+	{
+		if ($this->type() === 'site') {
+			return App::instance()->site();
+		}
+
+		if ($this->type() === 'user') {
+			return App::instance()->user($this->protocol->host());
+		}
+
+		if ($this->model = $this->findFromCache()) {
+			return $this->model;
+		}
+
+		$this->model = $this->findFromIndex();
+
+		if ($this->collection !== null) {
+			$this->model ??= $this->findFromIndex(true);
+		}
+
+		if ($this->model !== null) {
+			// lazily fill cache by writing UUID to cache
+			// whenever looked up from index
+			$this->populate();
+
+			return $this->model;
+		}
+
+		return null;
+	}
+
+	/**
+	 * Returns the model type
+	 *
+	 * @return string
+	 */
+	public function type(): string
+	{
+		return $this->protocol->type();
+	}
+
+	/**
+	 * Returns the UUID hash and type schema
+	 *
+	 * @return string
+	 */
+	public function uuid(): string
+	{
+		if ($this->protocol->host() === null) {
+			$this->create();
+		}
+
+		return $this->protocol->toString();
+	}
+
+	/**
+	 * Returns the value that will be stored in cache
+	 *
+	 * @return string
+	 */
+	public function value(): string
+	{
+		if (
+			$this->type() === 'site' ||
+			$this->type() === 'page' ||
+			$this->type() === 'user'
+		) {
+			return $this->model->id();
+		}
+
+		if ($this->type() === 'file') {
```

Comment:
Shouldn't the $file->id() method also always return the full id including the parent id?

Reply:
But we don't want the full id with parent id, we want parent uuid + filename, e.g. `page://page-uuid/filename.jpg` instead of just `page-a/filename.jpg`

Reply:
Forget my comment. I got confused by what value does.